In [1]:
import pandas as pd
import numpy as np

In [2]:
# original dataset
prev = pd.read_csv(r"C:\Users\ppo\OneDrive - San Diego Association of Governments\Shared Documents\Projects\2022\2022-87 School Inventory & Enrollment 2022 QC\data\inventory\QC_SCHOOL_2022.csv")
display(prev.head())
prev.shape

OID_       cdsCode                district                            name  \
0     1  3.768412e+13   Sweetwater Union High             Bonita Vista Middle   
1     2  3.768411e+13   Sweetwater Union High        Bonita Vista Senior High   
2     3  3.768024e+13  Chula Vista Elementary                      Clear View   
3     4  3.768024e+13  Chula Vista Elementary  Tiffany (Burton C.) Elementary   
4     5  3.768024e+13  Chula Vista Elementary     McMillin (Corky) Elementary   

                     addr         city    zip           openDate charter  \
0     650 Otay Lakes Road  Chula Vista  91910   7/1/1980 0:00:00       N   
1     751 Otay Lakes Road  Chula Vista  91913   7/1/1980 0:00:00       N   
2        455 Windrose Way  Chula Vista  91910  9/10/1991 0:00:00       N   
3    1691 Elmhurst Street  Chula Vista  91913   7/1/1980 0:00:00       N   
4  1201 Santa Cora Avenue  Chula Vista  91913  7/17/2001 0:00:00       N   

                      docType  ...          shrtName priv ClosedDate Phantom  \
0        High School District  ...  Bonita Vista Mid    N        NaN     0.0   
1        High School District  ...   Bonita Vista SH    N        NaN     0.0   
2  Elementary School District  ...        Clear View    N        NaN     0.0   
3  Elementary School District  ...           Tiffany    N        NaN     0.0   
4  Elementary School District  ...          McMillin    N        NaN     0.0   

       Notes  Check_ regionID  schoolID  Change    cdeCodeNUM  
0  No change     0.0      0.0       1.0     0.0  3.768412e+13  
1  No change     0.0      0.0       2.0     0.0  3.768411e+13  
2  No change     0.0      0.0       3.0     0.0  3.768024e+13  
3  No change     0.0      0.0       4.0     0.0  3.768024e+13  
4  No change     0.0      0.0       5.0     0.0  3.768024e+13  

[5 rows x 22 columns]

(1653, 22)

In [3]:
# the empty rows are deleted in revised
revised = pd.read_csv(r"C:\Users\ppo\Downloads\QC_SCHOOL_2022_Revised.csv")
display(revised.head())
revised.shape

OID_       cdsCode                district                            name  \
0     1  3.770000e+13   Sweetwater Union High             Bonita Vista Middle   
1     2  3.770000e+13   Sweetwater Union High        Bonita Vista Senior High   
2     3  3.770000e+13  Chula Vista Elementary                      Clear View   
3     4  3.770000e+13  Chula Vista Elementary  Tiffany (Burton C.) Elementary   
4     5  3.770000e+13  Chula Vista Elementary     McMillin (Corky) Elementary   

                     addr         city    zip        openDate charter  \
0     650 Otay Lakes Road  Chula Vista  91910   7/1/1980 0:00       N   
1     751 Otay Lakes Road  Chula Vista  91913   7/1/1980 0:00       N   
2        455 Windrose Way  Chula Vista  91910  9/10/1991 0:00       N   
3    1691 Elmhurst Street  Chula Vista  91913   7/1/1980 0:00       N   
4  1201 Santa Cora Avenue  Chula Vista  91913  7/17/2001 0:00       N   

                      docType  ...          shrtName priv ClosedDate Phantom  \
0        High School District  ...  Bonita Vista Mid    N        NaN       0   
1        High School District  ...   Bonita Vista SH    N        NaN       0   
2  Elementary School District  ...        Clear View    N        NaN       0   
3  Elementary School District  ...           Tiffany    N        NaN       0   
4  Elementary School District  ...          McMillin    N        NaN       0   

       Notes  Check_ regionID  schoolID  Change    cdeCodeNUM  
0  No change       0        0         1       0  3.770000e+13  
1  No change       0        0         2       0  3.770000e+13  
2  No change       0        0         3       0  3.770000e+13  
3  No change       0        0         4       0  3.770000e+13  
4  No change       0        0         5       0  3.770000e+13  

[5 rows x 22 columns]

(1650, 22)

In [4]:
# focusing on active schools
df = prev[(prev['Phantom']==0) & (pd.isnull(prev['ClosedDate']))]
display(df.head())
df.shape

OID_       cdsCode                district                            name  \
0     1  3.768412e+13   Sweetwater Union High             Bonita Vista Middle   
1     2  3.768411e+13   Sweetwater Union High        Bonita Vista Senior High   
2     3  3.768024e+13  Chula Vista Elementary                      Clear View   
3     4  3.768024e+13  Chula Vista Elementary  Tiffany (Burton C.) Elementary   
4     5  3.768024e+13  Chula Vista Elementary     McMillin (Corky) Elementary   

                     addr         city    zip           openDate charter  \
0     650 Otay Lakes Road  Chula Vista  91910   7/1/1980 0:00:00       N   
1     751 Otay Lakes Road  Chula Vista  91913   7/1/1980 0:00:00       N   
2        455 Windrose Way  Chula Vista  91910  9/10/1991 0:00:00       N   
3    1691 Elmhurst Street  Chula Vista  91913   7/1/1980 0:00:00       N   
4  1201 Santa Cora Avenue  Chula Vista  91913  7/17/2001 0:00:00       N   

                      docType  ...          shrtName priv ClosedDate Phantom  \
0        High School District  ...  Bonita Vista Mid    N        NaN     0.0   
1        High School District  ...   Bonita Vista SH    N        NaN     0.0   
2  Elementary School District  ...        Clear View    N        NaN     0.0   
3  Elementary School District  ...           Tiffany    N        NaN     0.0   
4  Elementary School District  ...          McMillin    N        NaN     0.0   

       Notes  Check_ regionID  schoolID  Change    cdeCodeNUM  
0  No change     0.0      0.0       1.0     0.0  3.768412e+13  
1  No change     0.0      0.0       2.0     0.0  3.768411e+13  
2  No change     0.0      0.0       3.0     0.0  3.768024e+13  
3  No change     0.0      0.0       4.0     0.0  3.768024e+13  
4  No change     0.0      0.0       5.0     0.0  3.768024e+13  

[5 rows x 22 columns]

(1034, 22)

In [5]:
prev[(prev['Phantom']==0) & (~pd.isnull(prev['ClosedDate']))]

OID_       cdsCode       district            name              addr  \
1649  1650  3.768297e+13  Poway Unified  Discovery Isle  11740 Creek Road   

       city    zip          openDate charter                  docType  ...  \
1649  Poway  92064  7/1/1980 0:00:00       N  Unified School District  ...   

               shrtName priv        ClosedDate Phantom  \
1649  Discovery Isle P2    Y  1/1/2022 0:00:00     0.0   

                                                  Notes  Check_ regionID  \
1649  PHO: 11/29/2022 - reopened, assigned new schoolID    11.0      0.0   

      schoolID  Change    cdeCodeNUM  
1649  100437.0     0.0  3.768297e+13  

[1 rows x 22 columns]

schoolID 100437 might be problematic since active schools should have null as closedDate

In [6]:
df['schoolID'].nunique()

1034

In [7]:
df.columns

Index(['OID_', 'cdsCode', 'district', 'name', 'addr', 'city', 'zip',
       'openDate', 'charter', 'docType', 'socType', 'gsOffered', 'shrtName',
       'priv', 'ClosedDate', 'Phantom', 'Notes', 'Check_', 'regionID',
       'schoolID', 'Change', 'cdeCodeNUM'],
      dtype='object')

In [8]:
df.dtypes

OID_            int64
cdsCode       float64
district       object
name           object
addr           object
city           object
zip            object
openDate       object
charter        object
docType        object
socType        object
gsOffered      object
shrtName       object
priv           object
ClosedDate     object
Phantom       float64
Notes          object
Check_        float64
regionID      float64
schoolID      float64
Change        float64
cdeCodeNUM    float64
dtype: object

Checking uniqueness:

In [9]:
unique_cols = ['schoolID']

In [10]:
def unique(cols):
    not_unique = []
    for col in cols:
        ser = df[col]
        if ser.nunique() != len(ser):
            not_unique.append(col)
    return not_unique

In [11]:
unique(unique_cols)

[]

Checking nulls:

In [12]:
null_cols = df.columns

In [13]:
def not_null(cols):
    nulls = []
    for col in cols:
        ser = df[col]
        if len(df[pd.isna(ser)]) > 0:
            nulls.append(col)
    return nulls

In [14]:
not_null(null_cols)

['ClosedDate', 'Notes']

Let's investigate how many null vals there are in these columns

In [15]:
for col in not_null(null_cols):
    # display(df[pd.isna(df[col])])
    print(f'{col} column has {len(df[pd.isna(df[col])])} missing row(s)')

ClosedDate column has 1034 missing row(s)
Notes column has 54 missing row(s)


Checking fixed numbers of vals:

In [16]:
fixed_cols = 'charter docType socType priv Phantom regionID'.split()

In [17]:
def fixed(cols):
    dict = {}
    for col in cols:
        ser = df[col]
        dict[col] = ser.unique().tolist()
    return dict

In [18]:
fixed(fixed_cols)

{'charter': ['N', 'Y'],
 'docType': ['High School District',
  'Elementary School District',
  'Unified School District',
  'County Office of Education (COE)',
  'Statewide Benefit Charter',
  'Community College District',
  'Regional Occupation Center/Program (ROC/P)',
  'State Board of Education'],
 'socType': ['Intermediate/Middle Schools (Public)',
  'High Schools (Public)',
  'Elementary Schools (Public)',
  'Alternative Schools of Choice',
  'Continuation High Schools',
  'District Community Day Schools',
  'K-12 Schools (Public)',
  'Special Education Schools (Public)',
  'Junior High Schools (Public)',
  'Elemen Schools In 1 School Dist. (Public)',
  'County Community',
  'Juvenile Court Schools',
  'Preschool',
  'Private',
  'Adult Education Centers',
  'ROC/ROP',
  'High Schools In 1 School Dist. (Public)',
  'San Marcos Unified'],
 'priv': ['N', 'Y'],
 'Phantom': [0.0],
 'regionID': [0.0]}

In [19]:
df['district'].value_counts()

San Diego Unified                                           304
Chula Vista Elementary                                       63
Poway Unified                                                54
Vista Unified                                                45
Sweetwater Union High                                        38
La Mesa-Spring Valley                                        36
Oceanside Unified                                            34
Escondido Union                                              33
Cajon Valley Union                                           32
Grossmont Union High                                         27
San Marcos Unified                                           26
Carlsbad Unified                                             25
Fallbrook Union Elementary                                   21
San Diego County Office of Education                         19
San Dieguito Union High                                      19
Mountain Empire Unified                 

In [20]:
df[df['city']=="Fallbrook "]

OID_       cdsCode                    district             name  \
1609  1610  3.768115e+13  Fallbrook Union Elementary  Topview Academy   

                  addr        city    zip           openDate charter  \
1609  3843 Canonita Dr  Fallbrook   92028  12/7/2021 0:00:00       N   

                         docType  ...         shrtName priv ClosedDate  \
1609  Elementary School District  ...  Topview Academy    Y        NaN   

     Phantom                  Notes  Check_ regionID  schoolID  Change  \
1609     0.0  PHO: Added 11/29/2022     0.0      0.0  100406.0     0.0   

        cdeCodeNUM  
1609  3.768115e+13  

[1 rows x 22 columns]

In [21]:
(df['district'].value_counts().index).sort_values()

Index(['Alpine Union Elementary', 'Bonsall Unified', 'Borrego Springs Unified',
       'Cajon Valley Union', 'Cajon Valley Union Elementary',
       'Cardiff Elementary', 'Carlsbad Unified', 'Chula Vista Elementary',
       'Coronado Unified', 'Dehesa Elementary', 'Del Mar Union Elementary',
       'Encinitas Union Elementary', 'Escondido Union',
       'Escondido Union Elementary', 'Escondido Union High',
       'Fallbrook Union Elementary', 'Fallbrook Union High',
       'Grossmont Union High', 'Imperial Beach Charter',
       'Jamul-Dulzura Union Elementary', 'Julian Union Elementary',
       'Julian Union High', 'La Mesa-Spring Valley',
       'Lakeside Union Elementary', 'Lemon Grove', 'Lemon Grove Elementary',
       'Mono County Office of Education', 'Mountain Empire Unified',
       'National Elementary', 'Oceanside Unified', 'Poway Unified',
       'Ramona City Unified', 'Ramona Unified', 'Rancho Santa Fe Elementary',
       'SBC - High Tech High', 'SBE - Baypoint Preparatory 

In [22]:
(df['city'].value_counts().index).sort_values()

Index(['Alpine', 'Bonita', 'Bonsall', 'Borrego Springs', 'Boulevard', 'Campo',
       'Cardiff-by-the-Sea', 'Carlsbad', 'Chula Vista', 'Coronado', 'Del Mar',
       'Descanso', 'El Cajon', 'El Cajon ', 'Encinitas', 'Escondido',
       'Escondido ', 'Fallbrook', 'Fallbrook ', 'Imperial Beach', 'Jamul',
       'Julian', 'La Jolla', 'La Jolla ', 'La Mesa', 'Lakeside', 'Lemon Grove',
       'National City', 'Oceanside', 'Pala', 'Pauma Valley', 'Pine Valley',
       'Potrero', 'Poway', 'Rainbow', 'Ramona', 'Rancho Santa Fe',
       'San Clemente', 'San Diego', 'San Diego ', 'San Marcos', 'San Ysidro',
       'Santa Ysabel', 'Santee', 'Solana Beach', 'Spring Valley', 'Tecate',
       'Valley Center', 'Vista', 'Warner Springs', 'redacted'],
      dtype='object')

Note that some of the entries for the column "city" includes a space at the end. Might not be a big issue when working in excel but generally best practice to fix it because they will be counted as different cities when performing analysis

In [23]:
(df['socType'].value_counts().index).sort_values()

Index(['Adult Education Centers', 'Alternative Schools of Choice',
       'Continuation High Schools', 'County Community',
       'District Community Day Schools',
       'Elemen Schools In 1 School Dist. (Public)',
       'Elementary Schools (Public)', 'High Schools (Public)',
       'High Schools In 1 School Dist. (Public)',
       'Intermediate/Middle Schools (Public)', 'Junior High Schools (Public)',
       'Juvenile Court Schools', 'K-12 Schools (Public)', 'Preschool',
       'Private', 'ROC/ROP', 'San Marcos Unified',
       'Special Education Schools (Public)'],
      dtype='object')

Additional checks

In [24]:
df['gsOffered'].unique()

array(['7-8', '9-12', 'K-6', 'K-5', 'K-8', 'P-8', 'K-12', '7-12', '6-12',
       'P-6', '6-8', '1-5', 'K-4', '11-12', 'K-3', '3-5', '5-6', '7-9',
       '4-6', 'P-5', 'K-2', '10-12', 'P-12', 'P-K', 'K-1', '3-6', 'P',
       '4-8', '1-12', '5-8', 'Adult', '1-9', '1-6', 'K', '12', 'K-11',
       ' ', '9-Adult', '3-12', '9', '12-Adult', 'K-9', '8-12', '1-8',
       '3-9', 'K-K'], dtype=object)

In [25]:
df['district'].value_counts().index

Index(['San Diego Unified', 'Chula Vista Elementary', 'Poway Unified',
       'Vista Unified', 'Sweetwater Union High', 'La Mesa-Spring Valley',
       'Oceanside Unified', 'Escondido Union', 'Cajon Valley Union',
       'Grossmont Union High', 'San Marcos Unified', 'Carlsbad Unified',
       'Fallbrook Union Elementary', 'San Diego County Office of Education',
       'San Dieguito Union High', 'Mountain Empire Unified',
       'National Elementary', 'Lakeside Union Elementary',
       'Encinitas Union Elementary', 'Valley Center-Pauma Unified',
       'Solana Beach Elementary', 'South Bay Union Elementary',
       'Ramona City Unified', 'Del Mar Union Elementary', 'Santee Elementary',
       'Escondido Union High', 'San Ysidro Elementary', 'Warner Unified',
       'Coronado Unified', 'Lemon Grove', 'SBC - High Tech High',
       'Julian Union Elementary', 'Dehesa Elementary', 'Bonsall Unified',
       'Cajon Valley Union Elementary', 'Alpine Union Elementary',
       'Escondido Union 

Looking at unique vals of schoolID from POLY table

In [26]:
test = pd.read_csv(r"C:\Users\ppo\OneDrive - San Diego Association of Governments\Shared Documents\Projects\2022\2022-87 School Inventory & Enrollment 2022 QC\data\inventory\QC_SCHOOL_POLY_2022.csv")
display(test.head())
print(test.shape)
test['schoolID'].nunique()

OID_  schoolID         cdsCode                  district  \
0     1        88  37683386040042         San Diego Unified   
1     2       475  37680566038111  Del Mar Union Elementary   
2     3       403  37683386040208         San Diego Unified   
3     4       511  37683386059687         San Diego Unified   
4     5       570  37683386039846         San Diego Unified   

                         name                    addr       city    zip  \
0    Pacific Beach Elementary  1234 Tourmaline Street  San Diego  92109   
1  Del Mar Heights Elementary     13555 Boquita Drive    Del Mar  92014   
2      Sunset View Elementary        4365 Hill Street  San Diego  92107   
3            Muirlands Middle    1056 Nautilus Street   La Jolla  92037   
4         La Jolla Elementary      1111 Marine Street   La Jolla  92037   

           openDate charter                     docType  \
0  7/1/1980 0:00:00       N     Unified School District   
1  7/1/1980 0:00:00       N  Elementary School District   
2  7/1/1980 0:00:00       N     Unified School District   
3  7/1/1980 0:00:00       N     Unified School District   
4  7/1/1980 0:00:00       N     Unified School District   

                                socType gsOffered            shrtName priv  \
0           Elementary Schools (Public)       K-5  Pacific Beach Elem    N   
1           Elementary Schools (Public)       K-6     Del Mar Heights    N   
2           Elementary Schools (Public)       K-4         Sunset View    N   
3  Intermediate/Middle Schools (Public)       6-8           Muirlands    N   
4           Elementary Schools (Public)       K-5       La Jolla Elem    N   

      Shape_STAr   Shape_STLe  Shape_Length     Shape_Area  
0  334304.483476  2472.448920   2472.448920  334304.483479  
1  475624.520306  2808.574191   2808.574191  475624.520309  
2  376604.128407  2398.947251   2398.947251  376604.128409  
3  744170.258130  3809.709403   3809.709403  744170.258131  
4  362896.936759  2549.731422   2549.731422  362896.936758

(1031, 19)


1031

In [27]:
df['schoolID'][~df['schoolID'].isin(test['schoolID'])]

1636    100433.0
1637    100434.0
1638    100435.0
1639    100436.0
Name: schoolID, dtype: float64

Noe that schoolIDs 100433, 100434, 100435, and 100436 exist in SCHOOL_2022 but not in POLY_2022

In [28]:
df[df['schoolID']==100433].T

1636
OID_                                                     1637
cdsCode                                            3.7681e+13
district                                      Escondido Union
name                                       Visitation Academy
addr                                                 redacted
city                                                 redacted
zip                                                  redacted
openDate                                    4/28/2022 0:00:00
charter                                                     N
docType                            Elementary School District
socType                                               Private
gsOffered                                                K-12
shrtName                                   Visitation Academy
priv                                                        Y
ClosedDate                                                NaN
Phantom                                                     0
Notes       PHO: Added 11/29/2022. Unable to geocode, reda...
Check_                                                      0
regionID                                                    0
schoolID                                               100433
Change                                                      0
cdeCodeNUM                                         3.7681e+13

Reason: no address provided

In [29]:
test.sort_values(by=['schoolID']).tail(5)

OID_  schoolID         cdsCode                     district  \
1006  1007    100431  37756146164198  Valley Center-Pauma Unified   
993    994    100432  37682966135149                Poway Unified   
1028  1029    100437  37682966160089                Poway Unified   
1030  1031    100438  37737916927644            Oceanside Unified   
1029  1030    100439  37683387080427            San Diego Unified   

                               name                                 addr  \
1006                  Arrow Academy                 13826 Misty Oak Road   
993       Carmel Mountain Preschool  9510 Carmel Mountain Preschool Road   
1028                 Discovery Isle                     11740 Creek Road   
1030  TERI, INC. The Country School                     3225 Roymar Road   
1029         Bright Horizon Academy                7202 Princess View Dr   

               city    zip            openDate charter  \
1006  Valley Center  92082   12/8/2021 0:00:00       N   
993       San Diego  92129  11/10/2008 0:00:00       N   
1028          Poway  92064    7/1/1980 0:00:00       N   
1030      Oceanside  92054    7/1/1980 0:00:00       N   
1029      San Diego  92120    7/1/1980 0:00:00       N   

                      docType             socType gsOffered  \
1006  Unified School District             Private      1-12   
993   Unified School District             Private         K   
1028  Unified School District             Private         K   
1030  Unified School District  San Marcos Unified      6-12   
1029  Unified School District             Private      K-12   

               shrtName priv     Shape_STAr   Shape_STLe  Shape_Length  \
1006      Arrow Academy    Y  100433.327139  1278.625336   1278.625336   
993                 CMP    Y  130065.369040  1791.960984   1791.960984   
1028  Discovery Isle P2    Y  105052.630261  1403.037652   1403.037652   
1030           TERI TCS    Y   30710.204100   703.321338    703.321338   
1029  Bright Horizon AG    Y  179103.686915  1738.406053   1738.406053   

         Shape_Area  
1006  100433.327139  
993   130065.369040  
1028  105052.630261  
1030   30710.204100  
1029  179103.686915

### No need to do the following because our df consists of schoolIDs that does not have closedDate
Previously used it for the dataframe "revised"

Verify that all closed dates are after open dates 

Many rows have missing CloseDate values so let's filter out those values

In [30]:
df = revised

In [31]:
dates = df[~pd.isna(df['ClosedDate'])]
dates

OID_       cdsCode                     district  \
15      16  3.770000e+13            Dehesa Elementary   
106    107  3.770000e+13           Cajon Valley Union   
131    132  3.770000e+13           San Marcos Unified   
168    169  3.770000e+13            San Diego Unified   
184    185  3.770000e+13        La Mesa-Spring Valley   
...    ...           ...                          ...   
1597  1598  3.770000e+13        San Ysidro Elementary   
1598  1599  3.770000e+13                Vista Unified   
1599  1600  3.770000e+13            Oceanside Unified   
1600  1601  3.780000e+13  Valley Center-Pauma Unified   
1647  1650  3.770000e+13                Poway Unified   

                              name                          addr  \
15                  Dehesa Charter  1441 Montiel Road, Suite 143   
106     Cajon Valley Community Day              165 Roanoke Road   
131   Bayshore Preparatory Charter        1175 Linda Vista Drive   
168                      Logan K-8     2875 Ocean View Boulevard   
184           Spring Valley Middle             3900 Conrad Drive   
...                            ...                           ...   
1597            Master Classes PSP      4184 Beyer Blvd Suit 201   
1598          Institute of Wisdom                      356 Crest   
1599             Oceans Elementary         509 Sportfisher Drive   
1600           Oney Private School             30451 Miller Road   
1647                Discovery Isle              11740 Creek Road   

               city    zip         openDate charter  \
15        Escondido  92026  10/15/2001 0:00       Y   
106        El Cajon  92021   8/30/1999 0:00       N   
131      San Marcos  92078    9/1/2005 0:00       Y   
168       San Diego  92113    7/1/1980 0:00       N   
184   Spring Valley  91977    7/1/1980 0:00       N   
...             ...    ...              ...     ...   
1597     San Ysidro  92173    1/7/2021 0:00       N   
1598          Vista  92084   1/15/2021 0:00       N   
1599      Oceanside  92054   1/15/2021 0:00       N   
1600  Valley Center  92082  10/14/2020 0:00       N   
1647          Poway  92064    7/1/1980 0:00       N   

                         docType  ...              shrtName priv  \
15    Elementary School District  ...        Dehesa Charter    N   
106   Elementary School District  ...      Cajon Valley Day    N   
131      Unified School District  ...         Bayshore Prep    N   
168      Unified School District  ...                 Logan    N   
184   Elementary School District  ...     Spring Valley Mid    N   
...                          ...  ...                   ...  ...   
1597  Elementary School District  ...    Master Classes PSP    Y   
1598     Unified School District  ...  Institute of Wisdom     Y   
1599     Unified School District  ...     Oceans Elementary    Y   
1600     Unified School District  ...                  Oney    Y   
1647     Unified School District  ...     Discovery Isle P2    Y   

          ClosedDate Phantom  \
15    6/30/2019 0:00       1   
106   6/16/2017 0:00       1   
131   6/30/2018 0:00       1   
168   6/30/2020 0:00       1   
184   6/30/2016 0:00       1   
...              ...     ...   
1597   7/1/2022 0:00       1   
1598   7/1/2022 0:00       1   
1599   7/1/2022 0:00       1   
1600   7/1/2022 0:00       1   
1647   1/1/2022 0:00       0   

                                                  Notes  Check_ regionID  \
15           Coincident with additional school point(s)       0        0   
106                                              Closed       0        0   
131                                              Closed       0        0   
168                                           No change       0        0   
184                                              Closed       0        0   
...                                                 ...     ...      ...   
1597                      PHO: Added 1/28/2022. Closed        0        0   
1598  PHO: Added 1/28/2022. 11/29

In [32]:
df.dtypes

OID_            int64
cdsCode       float64
district       object
name           object
addr           object
city           object
zip            object
openDate       object
charter        object
docType        object
socType        object
gsOffered      object
shrtName       object
priv           object
ClosedDate     object
Phantom         int64
Notes          object
Check_          int64
regionID        int64
schoolID        int64
Change          int64
cdeCodeNUM    float64
dtype: object

Lets change the openDate and ClosedDate attributes to datetime object so we can compare the two

In [33]:
dates['openDate'] = pd.to_datetime(dates['openDate'])
dates['ClosedDate'] = pd.to_datetime(dates['ClosedDate'])
dates.dtypes

c:\Users\ppo\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
c:\Users\ppo\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


OID_                   int64
cdsCode              float64
district              object
name                  object
addr                  object
city                  object
zip                   object
openDate      datetime64[ns]
charter               object
docType               object
socType               object
gsOffered             object
shrtName              object
priv                  object
ClosedDate    datetime64[ns]
Phantom                int64
Notes                 object
Check_                 int64
regionID               int64
schoolID               int64
Change                 int64
cdeCodeNUM           float64
dtype: object

In [34]:
dates.head()

OID_       cdsCode               district                          name  \
15     16  3.770000e+13      Dehesa Elementary                Dehesa Charter   
106   107  3.770000e+13     Cajon Valley Union    Cajon Valley Community Day   
131   132  3.770000e+13     San Marcos Unified  Bayshore Preparatory Charter   
168   169  3.770000e+13      San Diego Unified                     Logan K-8   
184   185  3.770000e+13  La Mesa-Spring Valley          Spring Valley Middle   

                             addr           city    zip   openDate charter  \
15   1441 Montiel Road, Suite 143      Escondido  92026 2001-10-15       Y   
106              165 Roanoke Road       El Cajon  92021 1999-08-30       N   
131        1175 Linda Vista Drive     San Marcos  92078 2005-09-01       Y   
168     2875 Ocean View Boulevard      San Diego  92113 1980-07-01       N   
184             3900 Conrad Drive  Spring Valley  91977 1980-07-01       N   

                        docType  ...           shrtName priv ClosedDate  \
15   Elementary School District  ...     Dehesa Charter    N 2019-06-30   
106  Elementary School District  ...   Cajon Valley Day    N 2017-06-16   
131     Unified School District  ...      Bayshore Prep    N 2018-06-30   
168     Unified School District  ...              Logan    N 2020-06-30   
184  Elementary School District  ...  Spring Valley Mid    N 2016-06-30   

    Phantom                                       Notes  Check_ regionID  \
15        1  Coincident with additional school point(s)       0        0   
106       1                                      Closed       0        0   
131       1                                      Closed       0        0   
168       1                                   No change       0        0   
184       1                                      Closed       0        0   

     schoolID  Change    cdeCodeNUM  
15         16       0  3.770000e+13  
106       107       0  3.770000e+13  
131       132       0  3.770000e+13  
168       169       0  3.770000e+13  
184       185       0  3.770000e+13  

[5 rows x 22 columns]

In [35]:
access_time = dates['ClosedDate'] - dates['openDate']
dates['access_time'] = access_time

c:\Users\ppo\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [36]:
dates[dates['access_time'].dt.days == 0]

OID_       cdsCode               district  \
1243  1244  3.770000e+13       Coronado Unified   
1286  1287  3.770000e+13  La Mesa-Spring Valley   
1293  1294  3.770000e+13            Lemon Grove   
1326  1327  3.770000e+13      San Diego Unified   
1328  1329  3.770000e+13      San Diego Unified   
1329  1330  3.770000e+13      San Diego Unified   
1338  1339  3.770000e+13      San Diego Unified   
1348  1349  3.770000e+13      San Diego Unified   
1358  1359  3.770000e+13      San Diego Unified   
1411  1412  3.770000e+13      San Diego Unified   
1474  1475  3.780000e+13         Warner Unified   

                                                   name  \
1243                        Coronado Village Elementary   
1286                           Spring Valley Elementary   
1293                           Lions Leadership Academy   
1326                         Keiller Leadership Academy   
1328                King/Chavez Primary Academy Charter   
1329          Memorial Academy of Learning & Technology   
1338  International High School of Science, Technolo...   
1348                            Parks (Rosa) Elementary   
1358                  Lewis and Clark Community Charter   
1411                            Global Learning Charter   
1474                                           Warner i   

                             addr            city    zip   openDate charter  \
1243             600 Sixth Street        Coronado  92118 1991-09-01       N   
1286            3845 Spring Drive   Spring Valley  91977 1980-07-01       N   
1293              8705 Echo Drive         La Mesa  91941 2002-09-03       Y   
1326           7270 Lisbon Street       San Diego  92114 2005-05-24       N   
1328              415 31st Street       San Diego  92102 2005-03-26       N   
1329            2850 Logan Avenue       San Diego  92113 2006-01-09       N   
1338           1867 Pentecost Way       San Diego  92105 2015-09-09       Y   
1348           4510 Landis Street       San Diego  92105 1997-09-03       N   
1358          1855 Hancock Street       San Diego  92110 2002-03-20       Y   
1411  402 W. Broadway, Suite 1000       San Diego  92101 2002-09-03       Y   
1474             30951 Highway 79  Warner Springs  92086 2013-09-02       Y   

                         docType  ... priv ClosedDate Phantom   Notes Check_  \
1243     Unified School District  ...    N 1991-09-01       1  Closed      0   
1286  Elementary School District  ...    N 1980-07-01       1  Closed      0   
1293  Elementary School District  ...    N 2002-09-03       1  Closed      0   
1326     Unified School District  ...    N 2005-05-24       1  Closed      0   
1328     Unified School District  ...    N 2005-03-26       1  Closed      0   
1329     Unified School District  ...    N 2006-01-09       1  Closed      0   
1338     Unified School District  ...    N 2015-09-09       1  Closed      0   
1348     Unified School District  ...    N 1997-09-03       1  Closed      0   
1358     Unified School District  ...    N 2002-03-20       1  Closed      0   
1411     Unified School District  ...    N 2002-09-03       1  Closed      5   
1474     Unified School District  ...    N 2013-09-02       1  Closed      0   

      regionID schoolID  Change    cdeCodeNUM  access_time  
1243         0      960       0  3.770000e+13       0 days  
1286         0     1003       0  3.770000e+13       0 days  
1293         0     1010       0  3.770000e+13       0 days  
1326         0     1043       0  3.770000e+13       0 days  
1328         0     1045       0  3.770000e+13       0 days  
1329         0     1046       0  3.770000e+13       0 days  
1338         0     1055       0  3.770000e+13       0 days  
1348         0     1065       0  3.770000e+13       0 days  
1358         0     1075       0  3.770000e+13       0 days  
1411         0     1128       0  3.770000e+13       0 days  
1474         0     1191       0  3.780000e+13       0 days  

[11 rows x 23 columns]

In [37]:
dates[dates['access_time'].dt.days < 0]

OID_       cdsCode                   district  \
1276  1277  3.770000e+13  Lakeside Union Elementary   
1296  1297  3.770000e+13    Mountain Empire Unified   
1316  1317  3.770000e+13          San Diego Unified   
1430  1431  3.770000e+13      Sweetwater Union High   

                                    name                          addr  \
1276  Eastern County Preparatory Academy          2410 Mount Elena Way   
1296           J.P. Western Charter High     3291 Buckman Springs Road   
1316             P.R.I.D.E. Charter High          5825 Imperial Avenue   
1430        Advanced Achievement Academy  3772 Balboa Terrace, Suite A   

             city    zip   openDate charter                     docType  ...  \
1276        Jamul  91935 2010-08-30       Y  Elementary School District  ...   
1296  Pine Valley  91962 2012-07-01       Y     Unified School District  ...   
1316    San Diego  92114 2008-09-02       Y     Unified School District  ...   
1430    San Diego  92117 2006-09-01       Y        High School District  ...   

     priv ClosedDate Phantom  \
1276    N 2009-09-17       1   
1296    N 2012-05-08       1   
1316    N 2008-04-08       1   
1430    N 2006-05-25       1   

                                                  Notes Check_  regionID  \
1276                                             Closed      0         0   
1296                                             Closed      0         0   
1316                                             Closed      0         0   
1430  RRI. Original record has an open date after th...      0         0   

     schoolID  Change    cdeCodeNUM  access_time  
1276      993       0  3.770000e+13    -347 days  
1296     1013       0  3.770000e+13     -54 days  
1316     1033       0  3.770000e+13    -147 days  
1430     1147       0  3.770000e+13     -99 days  

[4 rows x 23 columns]

These rows have open date after the close date